File: frontend/lib/management-governance/directors.ts
/**
 * Director master helpers for Management & Governance.
 */

import type { DirectorRecord, ManagementGovernancePayload } from '@/lib/schemas/management-governance';

export type DirectorCategoryCounts = {
  total: number;
  current: number;
  proposed: number;
  executive: number;
  nonExecutive: number;
  independent: number;
  nominee: number;
  promoter: number;
  women: number;
  resident: number;
};

export type DirectorshipCounts = {
  totalCurrent: number;
  currentPublicCompany: number;
};

export type DirectorDeletionValidation = {
  canDelete: boolean;
  dependencies: string[];
};

export function getDirectors(payload: ManagementGovernancePayload): DirectorRecord[] {
  return payload.directorsProfilesAppointmentsAndEligibility.directors;
}

export function getDirectorById(
  payload: ManagementGovernancePayload,
  directorId: string,
): DirectorRecord | undefined {
  return getDirectors(payload).find((director) => director.id === directorId);
}

function isCurrentDirector(director: DirectorRecord): boolean {
  return director.appointmentStatus === 'current' || director.appointmentStatus === '';
}

function isIndependentDirector(director: DirectorRecord): boolean {
  return (
    director.designation === 'independent-director' ||
    director.independentStatus === 'yes'
  );
}

function isWomanDirector(director: DirectorRecord): boolean {
  return director.gender === 'female';
}

function isResidentDirector(director: DirectorRecord): boolean {
  const residence = director.countryOfResidence.trim().toLowerCase();
  return residence === 'india' || residence === 'in';
}

export function countByCategory(payload: ManagementGovernancePayload): DirectorCategoryCounts {
  const directors = getDirectors(payload);
  const currentDirectors = directors.filter(isCurrentDirector);

  return {
    total: directors.length,
    current: currentDirectors.length,
    proposed: directors.filter((d) => d.appointmentStatus.startsWith('proposed')).length,
    executive: currentDirectors.filter((d) => d.executiveNonExecutive === 'executive').length,
    nonExecutive: currentDirectors.filter((d) => d.executiveNonExecutive === 'non-executive')
      .length,
    independent: currentDirectors.filter(isIndependentDirector).length,
    nominee: currentDirectors.filter(
      (d) => d.designation === 'nominee-director' || d.nomineeStatus === 'yes',
    ).length,
    promoter: currentDirectors.filter((d) => d.promoterStatus === 'yes').length,
    women: currentDirectors.filter(isWomanDirector).length,
    resident: currentDirectors.filter(isResidentDirector).length,
  };
}

export function computeDirectorshipCounts(director: DirectorRecord): DirectorshipCounts {
  const current = director.otherDirectorships.filter((d) => d.currentOrCeased === 'current');
  return {
    totalCurrent: current.length,
    currentPublicCompany: current.filter(
      (d) =>
        d.entityListingStatus === 'public-listed' || d.entityListingStatus === 'public-unlisted',
    ).length,
  };
}

export function computeAgeFromDob(dateOfBirth: string): string {
  if (!dateOfBirth.trim()) return '';
  const dob = new Date(dateOfBirth);
  if (Number.isNaN(dob.getTime())) return '';
  const today = new Date();
  let age = today.getFullYear() - dob.getFullYear();
  const monthDiff = today.getMonth() - dob.getMonth();
  if (monthDiff < 0 || (monthDiff === 0 && today.getDate() < dob.getDate())) {
    age -= 1;
  }
  return String(age);
}

export function validateDirectorDeletion(
  payload: ManagementGovernancePayload,
  directorId: string,
): DirectorDeletionValidation {
  const dependencies: string[] = [];
  const board = payload.boardStructureAndIpoGovernanceReadiness;

  if (board.leadership.chairmanDirectorId === directorId) {
    dependencies.push('Board leadership — Chairman');
  }
  if (board.leadership.managingDirectorDirectorId === directorId) {
    dependencies.push('Board leadership — Managing Director');
  }
  if (board.leadership.ceoDirectorId === directorId) {
    dependencies.push('Board leadership — CEO');
  }
  if (board.leadership.leadIndependentDirectorId === directorId) {
    dependencies.push('Board leadership — Lead Independent Director');
  }
  if (board.leadership.wholeTimeDirectorIds.includes(directorId)) {
    dependencies.push('Board leadership — Whole-Time Director');
  }
  if (board.ipoCommittee.chairpersonDirectorId === directorId) {
    dependencies.push('IPO Committee — Chairperson');
  }
  if (board.ipoCommittee.memberDirectorIds.includes(directorId)) {
    dependencies.push('IPO Committee — Member');
  }

  for (const committee of payload.boardCommitteesAndGovernanceBodies.committees) {
    if (committee.chairpersonDirectorId === directorId) {
      dependencies.push(`Committee "${committee.name || committee.committeeType}" — Chairperson`);
    }
    for (const member of committee.members) {
      if (member.directorId === directorId) {
        dependencies.push(`Committee "${committee.name || committee.committeeType}" — Member`);
      }
    }
  }

  for (const row of payload.remunerationServiceContractsEsopsAndBenefits.directorRemuneration) {
    if (row.directorId === directorId) dependencies.push('Director remuneration record');
  }
  for (const row of payload.remunerationServiceContractsEsopsAndBenefits.executiveAppointmentTerms) {
    if (row.directorId === directorId) dependencies.push('Executive appointment terms');
  }
  for (const row of payload.interestsConflictsAndManagementRelationships.directorOfferDocumentInterests) {
    if (row.directorId === directorId) dependencies.push('Director offer-document interests');
  }
  for (const row of payload.changesContinuityAndSuccession.boardChanges) {
    if (row.directorId === directorId) dependencies.push('Board change record');
  }

  const kmpLinked = payload.kmpSeniorManagementAndOrganisationStructure.kmpSmpRecords.some(
    (k) => k.linkedDirectorId === directorId,
  );
  if (kmpLinked) dependencies.push('Linked KMP/SMP record');

  return {
    canDelete: dependencies.length === 0,
    dependencies: [...new Set(dependencies)],
  };
}

File: backend/app/modules/management_governance/directors.py
"""Director master helpers — ports frontend directors.ts."""

from __future__ import annotations

from typing import Any


def get_directors(payload: dict[str, Any]) -> list[dict[str, Any]]:
    section = payload.get("directorsProfilesAppointmentsAndEligibility") or {}
    directors = section.get("directors") or []
    return [d for d in directors if isinstance(d, dict)]


def get_director_by_id(payload: dict[str, Any], director_id: str) -> dict[str, Any] | None:
    for director in get_directors(payload):
        if director.get("id") == director_id:
            return director
    return None


def _is_current_director(director: dict[str, Any]) -> bool:
    status = director.get("appointmentStatus") or ""
    return status in ("current", "")


def _is_independent_director(director: dict[str, Any]) -> bool:
    return (
        director.get("designation") == "independent-director"
        or director.get("independentStatus") == "yes"
    )


def _is_woman_director(director: dict[str, Any]) -> bool:
    return director.get("gender") == "female"


def _is_resident_director(director: dict[str, Any]) -> bool:
    residence = str(director.get("countryOfResidence") or "").strip().lower()
    return residence in ("india", "in")


def count_by_category(payload: dict[str, Any]) -> dict[str, int]:
    directors = get_directors(payload)
    current_directors = [d for d in directors if _is_current_director(d)]

    return {
        "total": len(directors),
        "current": len(current_directors),
        "proposed": len(
            [
                d
                for d in directors
                if str(d.get("appointmentStatus") or "").startswith("proposed")
            ]
        ),
        "executive": len(
            [d for d in current_directors if d.get("executiveNonExecutive") == "executive"]
        ),
        "nonExecutive": len(
            [d for d in current_directors if d.get("executiveNonExecutive") == "non-executive"]
        ),
        "independent": len([d for d in current_directors if _is_independent_director(d)]),
        "nominee": len(
            [
                d
                for d in current_directors
                if d.get("designation") == "nominee-director" or d.get("nomineeStatus") == "yes"
            ]
        ),
        "promoter": len(
            [d for d in current_directors if d.get("promoterStatus") == "yes"]
        ),
        "women": len([d for d in current_directors if _is_woman_director(d)]),
        "resident": len([d for d in current_directors if _is_resident_director(d)]),
    }


def compute_directorship_counts(director: dict[str, Any]) -> dict[str, int]:
    current = [
        d
        for d in (director.get("otherDirectorships") or [])
        if isinstance(d, dict) and d.get("currentOrCeased") == "current"
    ]
    return {
        "totalCurrent": len(current),
        "currentPublicCompany": len(
            [
                d
                for d in current
                if d.get("entityListingStatus") in ("public-listed", "public-unlisted")
            ]
        ),
    }


def validate_director_deletion(payload: dict[str, Any], director_id: str) -> dict[str, Any]:
    dependencies: list[str] = []
    board = payload.get("boardStructureAndIpoGovernanceReadiness") or {}
    leadership = board.get("leadership") or {}

    if leadership.get("chairmanDirectorId") == director_id:
        dependencies.append("Board leadership — Chairman")
    if leadership.get("managingDirectorDirectorId") == director_id:
        dependencies.append("Board leadership — Managing Director")
    if leadership.get("ceoDirectorId") == director_id:
        dependencies.append("Board leadership — CEO")
    if leadership.get("leadIndependentDirectorId") == director_id:
        dependencies.append("Board leadership — Lead Independent Director")
    if director_id in (leadership.get("wholeTimeDirectorIds") or []):
        dependencies.append("Board leadership — Whole-Time Director")

    ipo_committee = board.get("ipoCommittee") or {}
    if ipo_committee.get("chairpersonDirectorId") == director_id:
        dependencies.append("IPO Committee — Chairperson")
    if director_id in (ipo_committee.get("memberDirectorIds") or []):
        dependencies.append("IPO Committee — Member")

    committees_section = payload.get("boardCommitteesAndGovernanceBodies") or {}
    for committee in committees_section.get("committees") or []:
        if not isinstance(committee, dict):
            continue
        label = committee.get("name") or committee.get("committeeType") or "Committee"
        if committee.get("chairpersonDirectorId") == director_id:
            dependencies.append(f'Committee "{label}" — Chairperson')
        for member in committee.get("members") or []:
            if isinstance(member, dict) and member.get("directorId") == director_id:
                dependencies.append(f'Committee "{label}" — Member')

    remuneration = payload.get("remunerationServiceContractsEsopsAndBenefits") or {}
    for row in remuneration.get("directorRemuneration") or []:
        if isinstance(row, dict) and row.get("directorId") == director_id:
            dependencies.append("Director remuneration record")
    for row in remuneration.get("executiveAppointmentTerms") or []:
        if isinstance(row, dict) and row.get("directorId") == director_id:
            dependencies.append("Executive appointment terms")

    interests = payload.get("interestsConflictsAndManagementRelationships") or {}
    for row in interests.get("directorOfferDocumentInterests") or []:
        if isinstance(row, dict) and row.get("directorId") == director_id:
            dependencies.append("Director offer-document interests")

    changes = payload.get("changesContinuityAndSuccession") or {}
    for row in changes.get("boardChanges") or []:
        if isinstance(row, dict) and row.get("directorId") == director_id:
            dependencies.append("Board change record")

    kmp_section = payload.get("kmpSeniorManagementAndOrganisationStructure") or {}
    if any(
        isinstance(k, dict) and k.get("linkedDirectorId") == director_id
        for k in kmp_section.get("kmpSmpRecords") or []
    ):
        dependencies.append("Linked KMP/SMP record")

    unique = list(dict.fromkeys(dependencies))
    return {"canDelete": len(unique) == 0, "dependencies": unique}
